### PR TITLE
AVRO-3416: Benchmarking project for C#

### PR DESCRIPTION
**Pull request by zcsizmadia**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-3416)
  - https://issues.apache.org/jira/browse/AVRO-3416
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by zcsizmadia** [comment]
Here is an output after running `dotnet run -c Release -f net6.0 --runtimes netcoreapp3.1 net5.0 net6.0`. It compares the performance using the supported target frameworks:

```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1566 (21H2)
Intel Xeon E-2276M CPU 2.80GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.200
  [Host]     : .NET 6.0.2 (6.0.222.6406), X64 RyuJIT
  Job-IPEIUO : .NET 5.0.14 (5.0.1422.5710), X64 RyuJIT
  Job-RRGFRG : .NET 6.0.2 (6.0.222.6406), X64 RyuJIT
  Job-GTOEYQ : .NET Core 3.1.18 (CoreCLR 4.700.21.35901, CoreFX 4.700.21.36305), X64 RyuJIT


|                     Method |        Job |       Runtime |     Toolchain |           Mean |         Error |        StdDev |         Median | Ratio | RatioSD |
|--------------------------- |----------- |-------------- |-------------- |---------------:|--------------:|--------------:|---------------:|------:|--------:|
|           ParseSchemaSmall | Job-IPEIUO |      .NET 5.0 |        net5.0 |   5,368.123 ns |   106.0116 ns |   171.1890 ns |   5,376.043 ns |  0.89 |    0.04 |
|           ParseSchemaSmall | Job-RRGFRG |      .NET 6.0 |        net6.0 |   5,260.316 ns |   102.5021 ns |   159.5834 ns |   5,270.131 ns |  0.88 |    0.03 |
|           ParseSchemaSmall | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |   5,991.246 ns |   119.2506 ns |   132.5467 ns |   5,939.753 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|             ParseSchemaBig | Job-IPEIUO |      .NET 5.0 |        net5.0 |  27,213.034 ns |   542.4003 ns |   742.4437 ns |  27,320.763 ns |  0.90 |    0.03 |
|             ParseSchemaBig | Job-RRGFRG |      .NET 6.0 |        net6.0 |  25,537.003 ns |   510.3941 ns |   698.6332 ns |  25,594.539 ns |  0.84 |    0.03 |
|             ParseSchemaBig | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |  30,305.242 ns |   591.4515 ns |   704.0809 ns |  29,959.442 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|   CreateGenericRecordSmall | Job-IPEIUO |      .NET 5.0 |        net5.0 |      35.563 ns |     0.7632 ns |     1.3367 ns |      35.377 ns |  0.93 |    0.04 |
|   CreateGenericRecordSmall | Job-RRGFRG |      .NET 6.0 |        net6.0 |      34.402 ns |     0.7424 ns |     1.3387 ns |      34.504 ns |  0.91 |    0.05 |
|   CreateGenericRecordSmall | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |      38.286 ns |     0.8324 ns |     0.9909 ns |      38.267 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|     CreateGenericRecordBig | Job-IPEIUO |      .NET 5.0 |        net5.0 |     269.117 ns |     5.4419 ns |     7.8046 ns |     269.361 ns |  0.90 |    0.03 |
|     CreateGenericRecordBig | Job-RRGFRG |      .NET 6.0 |        net6.0 |     254.519 ns |     5.0770 ns |     8.8920 ns |     253.981 ns |  0.85 |    0.03 |
|     CreateGenericRecordBig | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |     296.601 ns |     5.9758 ns |     6.6420 ns |     297.056 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|  CreateSpecificRecordSmall | Job-IPEIUO |      .NET 5.0 |        net5.0 |       3.415 ns |     0.1308 ns |     0.1917 ns |       3.419 ns |  1.03 |    0.09 |
|  CreateSpecificRecordSmall | Job-RRGFRG |      .NET 6.0 |        net6.0 |       3.502 ns |     0.1350 ns |     0.2991 ns |       3.394 ns |  1.05 |    0.11 |
|  CreateSpecificRecordSmall | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |       3.335 ns |     0.1263 ns |     0.2003 ns |       3.364 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|    CreateSpecificRecordBig | Job-IPEIUO |      .NET 5.0 |        net5.0 |      19.113 ns |     0.4324 ns |     0.4979 ns |      19.005 ns |  0.90 |    0.04 |
|    CreateSpecificRecordBig | Job-RRGFRG |      .NET 6.0 |        net6.0 |      20.124 ns |     0.4730 ns |     0.4645 ns |      20.002 ns |  0.95 |    0.05 |
|    CreateSpecificRecordBig | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |      20.960 ns |     0.4831 ns |     0.8333 ns |      20.903 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|  GenericRecordsToAvroSmall | Job-IPEIUO |      .NET 5.0 |        net5.0 |   6,113.914 ns |   122.2540 ns |   210.8822 ns |   6,169.239 ns |  0.85 |    0.03 |
|  GenericRecordsToAvroSmall | Job-RRGFRG |      .NET 6.0 |        net6.0 |   6,815.491 ns |   227.2785 ns |   644.7512 ns |   6,582.915 ns |  1.02 |    0.09 |
|  GenericRecordsToAvroSmall | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |   7,159.197 ns |   141.5107 ns |   211.8066 ns |   7,100.346 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|    GenericRecordsToAvroBig | Job-IPEIUO |      .NET 5.0 |        net5.0 |  52,900.343 ns | 1,054.8385 ns | 1,673.0857 ns |  52,738.501 ns |  0.85 |    0.04 |
|    GenericRecordsToAvroBig | Job-RRGFRG |      .NET 6.0 |        net6.0 |  59,350.784 ns | 1,124.9646 ns | 2,801.5523 ns |  59,078.088 ns |  0.97 |    0.07 |
|    GenericRecordsToAvroBig | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |  62,483.008 ns | 1,243.2020 ns | 2,241.7496 ns |  62,260.767 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|  AvroToGenericRecordsSmall | Job-IPEIUO |      .NET 5.0 |        net5.0 |  12,625.160 ns |   251.3151 ns |   368.3744 ns |  12,472.104 ns |  0.80 |    0.04 |
|  AvroToGenericRecordsSmall | Job-RRGFRG |      .NET 6.0 |        net6.0 |  11,786.542 ns |   229.6952 ns |   390.0399 ns |  11,866.382 ns |  0.75 |    0.04 |
|  AvroToGenericRecordsSmall | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |  15,759.458 ns |   309.0975 ns |   602.8710 ns |  15,704.080 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|    AvroToGenericRecordsBig | Job-IPEIUO |      .NET 5.0 |        net5.0 |  87,675.306 ns | 1,713.3323 ns | 2,457.2100 ns |  88,638.483 ns |  0.81 |    0.03 |
|    AvroToGenericRecordsBig | Job-RRGFRG |      .NET 6.0 |        net6.0 |  91,820.714 ns | 2,070.5412 ns | 6,007.0153 ns |  91,008.777 ns |  0.84 |    0.06 |
|    AvroToGenericRecordsBig | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 | 108,870.422 ns | 2,168.7417 ns | 3,110.3445 ns | 108,441.388 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
| SpecificRecordsToAvroSmall | Job-IPEIUO |      .NET 5.0 |        net5.0 |   6,321.069 ns |   120.9818 ns |   107.2472 ns |   6,292.596 ns |  0.75 |    0.02 |
| SpecificRecordsToAvroSmall | Job-RRGFRG |      .NET 6.0 |        net6.0 |   7,483.118 ns |   187.3997 ns |   540.6911 ns |   7,342.720 ns |  0.88 |    0.07 |
| SpecificRecordsToAvroSmall | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |   8,422.971 ns |   166.7530 ns |   287.6407 ns |   8,383.934 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|   SpecificRecordsToAvroBig | Job-IPEIUO |      .NET 5.0 |        net5.0 |  57,909.991 ns | 1,139.8303 ns | 1,399.8140 ns |  58,028.305 ns |  0.73 |    0.03 |
|   SpecificRecordsToAvroBig | Job-RRGFRG |      .NET 6.0 |        net6.0 |  64,307.557 ns | 1,276.7085 ns | 2,802.4061 ns |  64,117.413 ns |  0.81 |    0.05 |
|   SpecificRecordsToAvroBig | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |  79,076.464 ns | 1,565.4951 ns | 3,053.3780 ns |  77,922.321 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
| AvroToSpecificRecordsSmall | Job-IPEIUO |      .NET 5.0 |        net5.0 |  18,082.362 ns |   345.7132 ns |   449.5248 ns |  18,001.993 ns |  0.83 |    0.04 |
| AvroToSpecificRecordsSmall | Job-RRGFRG |      .NET 6.0 |        net6.0 |  15,506.962 ns |   306.5587 ns |   704.3696 ns |  15,432.040 ns |  0.72 |    0.04 |
| AvroToSpecificRecordsSmall | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 |  21,746.739 ns |   433.9476 ns |   793.4977 ns |  21,534.998 ns |  1.00 |    0.00 |
|                            |            |               |               |                |               |               |                |       |         |
|   AvroToSpecificRecordsBig | Job-IPEIUO |      .NET 5.0 |        net5.0 |  89,881.408 ns | 1,727.5581 ns | 2,056.5350 ns |  88,850.732 ns |  0.82 |    0.04 |
|   AvroToSpecificRecordsBig | Job-RRGFRG |      .NET 6.0 |        net6.0 |  84,335.490 ns | 1,676.3619 ns | 2,404.1881 ns |  83,939.539 ns |  0.77 |    0.03 |
|   AvroToSpecificRecordsBig | Job-GTOEYQ | .NET Core 3.1 | netcoreapp3.1 | 110,031.915 ns | 2,182.0629 ns | 4,457.3762 ns | 109,291.528 ns |  1.00 |    0.00 |
```

**Comment by KyleSchoonover** [comment]
This is neat, but maybe you can explain the differences between this and the perf project.  Is this meant to be a replacement for the perf project?  Also is the plan that this be added to the CI/CD or just run locally.  It would be a good idea to add documentation on how to run locally since there are additional steps that may need explanation.  Maybe a breakdown of the output and what the expectation is.  Are we just reporting the performance and not taking any action?

File: lang/csharp/src/apache/benchmark/com/benchmark/big/mailing_address.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ * limitations under the License.
+ */
+// ------------------------------------------------------------------------------
+// <auto-generated>
+//    Generated by avrogen, version 1.11.0.0
+//    Changes to this file may cause incorrect behavior and will be lost if code
+//    is regenerated
+// </auto-generated>
+// ------------------------------------------------------------------------------
+namespace com.benchmark.big
```
Comment (martin-g) [inline lang/csharp/src/apache/benchmark/com/benchmark/big/mailing_address.cs:25]: `org.apache.avro.benchmark.big` ?

**Review by martin-g — APPROVED** [review]

**Comment by KyleSchoonover** [comment]
As long as we get a report from the CI, I'm good with it.  I brought it down locally and got similar results.  It would be good to see this continually run as we merge.

**Comment by KyleSchoonover** [comment]
Although can you add a MD file to documents it's intent and use?  I know after it runs it spits out some explanation of what the results mean, but it would be a good copy and paste from that.

**Comment by zcsizmadia** [comment]
Running this from CI might take too much time

**Comment by martin-g** [comment]
I also don't think we should add it to CI workflow.
CI agents run on VMs which do not have guaranteed performance!

**Pull request merged**